src/app/EventManagement.cpp

```diff
@@ -0,0 +1,867 @@
+                    // element always succeeds.
+                    // Since we're calling CopyElement and we've checked
+                    // that there is space in the next buffer, we don't expect
+                    // this to fail.
+                    err = CopyToNextBuffer(eventBuffer);
+                    SuccessOrExit(err);
+                    // success; evict head unconditionally
+                    eventBuffer->mProcessEvictedElement = nullptr;
+                    err                                 = eventBuffer->EvictHead();
+                    // if unconditional eviction failed, this
+                    // means that we have no way of further
+                    // clearing the buffer.  fail out and let the
+                    // caller know that we could not honor the
+                    // request
+                    SuccessOrExit(err);
+                    continue;
+                }
+                // we cannot copy event outright. We remember the
+                // current required space in mAppData, we note the
+                // space requirements for the event in the current
+                // buffer and make that space in the next buffer.
+                eventBuffer->mRequiredSpacedForEvicted = requiredSpace;
+                eventBuffer                            = eventBuffer->GetNextCircularEventBuffer();
+
+                // Sanity check: return error  here on null event buffer.  If
+                // eventBuffer->mpNext were null, then the `EvictBuffer`
+                // would have succeeded -- the event was
+                // already in the final buffer.
+                VerifyOrExit(eventBuffer != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+                requiredSpace = ctx.mSpaceNeededForMovedEvent;
+            }
+        }
+        else
+        {
+            if (eventBuffer == mpEventBuffer)
+                break;
+            eventBuffer   = eventBuffer->GetPreviousCircularEventBuffer();
+            requiredSpace = eventBuffer->mRequiredSpacedForEvicted;
+            err           = CHIP_NO_ERROR;
+        }
+    }
+
+    // On exit, configure the top-level s.t. it will always fail to evict an element
+    mpEventBuffer->mProcessEvictedElement = AlwaysFail;
+    mpEventBuffer->mAppData               = nullptr;
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+CHIP_ERROR EventManagement::CalculateEventSize(EventLoggingDelegate * apDelegate, const EventOptions * apOptions,
+                                               uint32_t & requiredSize)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVWriter writer;
+    EventLoadOutContext ctxt       = EventLoadOutContext(writer, apOptions->mpEventSchema->mPriority,
+                                                   GetPriorityBuffer(apOptions->mpEventSchema->mPriority)->mLastEventNumber);
+    System::PacketBufferHandle buf = System::PacketBufferHandle::New(kMaxEventSizeReserve);
```

Comment:
nit/followup: most other systems (e.g. protobufs or even our CHIP endian writers) have an option to have a 'null output' that does not write anything and are just used to compute sizes.

Requiring an allocation of  a packet buffer handle just to do a size computation seems wasteful. TLV should have the option to 'compute size' without an underlying buffer.